compiler/rustc_arena/src/lib.rs

```diff
@@ -516,6 +517,8 @@ impl DroplessArena {
 
     #[inline]
     pub fn alloc_from_iter<T, I: IntoIterator<Item = T>>(&self, iter: I) -> &mut [T] {
+        // Warning: this function is reentrant: `iter` could hold a reference to `&self` and add
+        // elements to it during iteration.
```

Comment:
This is interesting. I'd never considered reentrancy here.

Reply:
I didn't either until I tried to change it and it exploded ^^ It's interesting, the TypedArena on crates.io forbids it by keeping its RefCell borrowed throughout iteration. I wonder if it would be worth it to forbid it too.

Reply:
AST->HIR lowering uses this reentrancy, a lot.